### PR TITLE
Combine CDN in a Box Dockerfiles using build stage targets where possible

### PR DESCRIPTION
**Pull request by zrhoffman**
## What does this PR (Pull Request) do?

Currently, if you try to build the `trafficmonitor-debug`, `trafficops-debug`, and `trafficstats-debug` images without the `trafficmonitor`, `trafficops`, and `trafficstats` existing,
```shell
docker-compose -f docker-compose.yml -f optional/docker-compose.debugging.yml build --parallel trafficops
```
it will fail to build with an error like
```dockerfile
 ---> 82cad2818884
Step 5/6 : FROM trafficops
pull access denied for trafficops, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
Service 'trafficops' failed to build
```
because you need to build those images first. You can get around that by simply attempting to build everything twice:
```
docker-compose -f docker-compose.yml -f optional/docker-compose.debugging.yml build --parallel
docker-compose -f docker-compose.yml -f optional/docker-compose.debugging.yml build --parallel
```
But updating RPMs in the debug images is still annoying because it requires building everything twice in this way. This PR fixes that behavior.

This PR
- Makes the Traffic Monitor, Traffic Ops, and Traffic Ops debug images build stage targets within the non-debug Dockerfiles
- Removes the `trafficmonitor-nondebug`, `trafficops-nondebug`, and `trafficstats-nondebug` services from the `docker-compose.debugging.yml` compose file
- Makes `edge` build the `edge` target of `cache/Dockerfile` and makes `mid-01` and `mid-02` build the `mid` target of `cache/Dockerfile`
- Adds the ability to debug the CDN in a Box enroller

- [x] This PR is not related to any Issue 

## Which Traffic Control components are affected by this PR?

- CDN in a Box

## What is the best way to verify this PR?

- Untag the `trafficmonitor`, `trafficops`, and `trafficstats` images and then verify that running
  ```shell
  docker-compose -f docker-compose.yml -f optional/docker-compose.debugging.yml build --parallel
  ```
  succeeds

- Following the debugging documentation, set a breakpoint in `enroller.go` and try to hit it

## The following criteria are ALL met by this PR

- [x] CiaB changed only, tests unnecessary
- [x] This PR contains documentation
- [x] An update to CHANGELOG.md is unnecessary
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

**Review by TaylorCFrey — COMMENTED** [review]
Here's a few thoughts and notes based on reading the changes. I'm working on validating functionality separately.

File: traffic_ops/testing/api/conf/traffic-ops-test.conf [hunk 1]
```diff
@@ -8,13 +8,13 @@
             "warning": "stdout"
         },
         "session": {
-            "timeoutInSecs": 60
+            "timeoutInSecs": 9999999
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply traffic_ops/testing/api/conf/traffic-ops-test.conf]: Didn't mean to commit this, reverted in 27c7bd3c26.

File: traffic_ops/testing/api/conf/traffic-ops-test.conf [hunk 2]
```diff
@@ -27,7 +27,7 @@
         }
     },
     "trafficOpsDB": {
-        "dbname": "to_test",
+        "dbname": "traffic_ops",
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply traffic_ops/testing/api/conf/traffic-ops-test.conf]: Didn't mean to commit this, either, reverted in 27c7bd3c26.

File: traffic_ops/testing/api/conf/traffic-ops-test.conf [hunk 3]
```diff
@@ -8,13 +8,13 @@
...
         "session": {
-            "timeoutInSecs": 60
+            "timeoutInSecs": 9999999
         },
         "includeSystemTests": false
     },
     "use_ims": true,
     "trafficOps": {
-        "URL": "https://localhost:8443",
+        "URL": "https://localhost:6443",
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply traffic_ops/testing/api/conf/traffic-ops-test.conf]: Also didn't mean to commit this, reverted in 27c7bd3c26.

File: docs/source/development/debugging.rst [hunk 4]
```diff
@@ -24,11 +24,49 @@ Debugging inside CDN-in-a-Box
...
 
+Enroller
+========
+
+* Stop CDN-in-a-Box if it is running and remove any existing volumes. Build/rebuild the ``enroller-debug`` image each time you have changed :atc-file:`infrastructure/cdn-in-a-box/enroller/enroller.go`. Then, start CDN-in-a-Box.
+
+.. code-block:: shell
+	:caption: docker-compose command for debugging the CDN in a Box Enroller
+
+	alias mydc='docker-compose -f docker-compose.yml -f docker-compose.expose-ports.yml optional/docker-compose.debugging.yml'
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply docs/source/development/debugging.rst]: `-f` added in e7bd6e0c18.

File: docs/source/development/debugging.rst [hunk 5]
```diff
@@ -44,14 +82,14 @@ Traffic Monitor
 
 * Still in ``infrastructure/cdn-in-a-box``, open ``variables.env`` and set ``TM_DEBUG_ENABLE`` to ``true``.
 
-* Stop CDN-in-a-Box if it is running and remove any existing volumes. Rebuild the ``trafficmonitor`` image to make sure it uses our fresh ``traffic_monitor.rpm``. Then, start CDN-in-a-Box.
+* Stop CDN-in-a-Box if it is running and remove any existing volumes. Build the ``trafficmonitor-debug`` image to make sure it uses our fresh ``traffic_monitor.rpm``. Then, start CDN-in-a-Box:
 
 .. code-block:: shell
 	:caption: docker-compose command for debugging Traffic Monitor
 
 	alias mydc='docker-compose -f docker-compose.yml -f docker-compose.expose-ports.yml optional/docker-compose.debugging.yml'
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply docs/source/development/debugging.rst]: `-f` added in e7bd6e0c18.

File: docs/source/development/debugging.rst [hunk 6]
```diff
@@ -93,14 +131,14 @@ Traffic Ops
 
 * Still in ``infrastructure/cdn-in-a-box``, open ``variables.env`` and set ``TO_DEBUG_ENABLE`` to ``true``.
 
-* Stop CDN-in-a-Box if it is running and remove any existing volumes. Rebuild the ``trafficops-go`` image to make sure it uses our fresh ``traffic_ops.rpm``. Then, start CDN-in-a-Box.
+* Stop CDN-in-a-Box if it is running and remove any existing volumes. Build the ``trafficops-debug`` image to make sure it uses our fresh ``traffic_ops.rpm``. Then, start CDN-in-a-Box:
 
 .. code-block:: shell
 	:caption: docker-compose command for debugging Traffic Ops
 
 	alias mydc='docker-compose -f docker-compose.yml -f docker-compose.expose-ports.yml optional/docker-compose.debugging.yml'
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply docs/source/development/debugging.rst]: `-f` added in e7bd6e0c18.

File: docs/source/development/debugging.rst [hunk 7]
```diff
@@ -241,14 +279,14 @@ Traffic Stats
 
 * Still in ``infrastructure/cdn-in-a-box``, open ``variables.env`` and set ``TS_DEBUG_ENABLE`` to ``true``.
 
-* Stop CDN-in-a-Box if it is running and remove any existing volumes. Rebuild the ``trafficstats`` image to make sure it uses our fresh ``traffic_stats.rpm``. Then, start CDN-in-a-Box.
+* Stop CDN-in-a-Box if it is running and remove any existing volumes. Build the ``trafficstats-debug`` image to make sure it uses our fresh ``traffic_stats.rpm``. Then, start CDN-in-a-Box:
 
 .. code-block:: shell
 	:caption: docker-compose command for debugging Traffic Stats
 
 	alias mydc='docker-compose -f docker-compose.yml -f docker-compose.expose-ports.yml optional/docker-compose.debugging.yml'
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply docs/source/development/debugging.rst]: `-f` added in e7bd6e0c18.

File: docs/source/development/debugging.rst [hunk 8]
```diff
@@ -24,11 +24,49 @@ Debugging inside CDN-in-a-Box
...
+				"request": "attach",
+				"mode": "remote",
+				"port": 2343,
+				"cwd": "${workspaceRoot}/",
+				"remotePath": "/tmp/go/src/github.com/apache/trafficcontrol/",
+			}
+		]
+	}
+
+* Use the debugging configuration you created to start debugging Traffic Monitor. It should connect without first breaking at any line.
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply docs/source/development/debugging.rst]: Changed to `Enroller` in c63d9b81bc.

File: docs/source/development/debugging.rst [hunk 9]
```diff
@@ -24,11 +24,49 @@ Debugging inside CDN-in-a-Box
...
+
+	alias mydc='docker-compose -f docker-compose.yml -f docker-compose.expose-ports.yml optional/docker-compose.debugging.yml'
+	mydc down -v
+	mydc build enroller
+	mydc up
+
+* Install `an IDE that supports delve <https://github.com/go-delve/delve/blob/master/Documentation/EditorIntegration.md>`_ and create a debugging configuration over port 2343. If you are using VS Code, the configuration should look like this:
+
+.. code-block:: json
+	:caption: VS Code launch.json for debugging the CDN in a Box Enrolle
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply docs/source/development/debugging.rst]: Added an *r* in 6d607a415c.

File: docs/source/development/debugging.rst [hunk 10]
```diff
@@ -24,11 +24,49 @@ Debugging inside CDN-in-a-Box
...
 * `Traffic Monitor`_
 * `Traffic Ops`_
 * `Traffic Router`_
 * `Traffic Stats`_
 
+Enroller
+========
+
+* Stop CDN-in-a-Box if it is running and remove any existing volumes. Build/rebuild the ``enroller-debug`` image each time you have changed :atc-file:`infrastructure/cdn-in-a-box/enroller/enroller.go`. Then, start CDN-in-a-Box.
+
```
[the remark this answers is not in the document]
Comment (zrhoffman) [reply docs/source/development/debugging.rst:39]: *Variable* line added in 7ef06ae388.

**Comment by TaylorCFrey** [comment]
Confirmed debugging possible in Goland and VSCode.

**Review by TaylorCFrey — APPROVED** [review]
Code reviewed and functionality tested. LGTM!

Approved!

**Pull request merged**